File: golang/2337.move-pieces-to-obtain-a-string.go

func canChange(start string, target string) bool {
	m := len(start)

	for i, j := 0, 0; i < m || j < m; i, j = i+1, j+1 {
		for i < m && start[i] == '_' {
			i++
		}

		for j < m && target[j] == '_' {
			j++
		}

		if i == m || j == m {
			return i == m && j == m
		}

		if start[i] != target[j] {
			return false
		}

		if start[i] == 'L' && i < j {
			return false
		}

		if start[i] == 'R' && i > j {
			return false
		}

	}

	return true
}

File: python/2337.move-pieces-to-obtain-a-string.py


class Solution:
    def canChange(self, start: str, target: str) -> bool:

        # L >=
        # R <=

        left, right = 0, 0

        for s, t in zip(start, target):
            # open
            if s == "R":
                if left > 0:
                    return False
                right += 1

            # open
            if t == "L":
                if right > 0:
                    return False
                left += 1

            # close
            if t == "R":
                if right == 0:
                    return False
                right -= 1

            # close
            if s == "L":
                if left == 0:
                    return False
                left -= 1

        return left == 0 and right == 0

